### PR TITLE
egl: add support for libglvnd

### PR DESCRIPTION
**Pull request by peat-psuwit**
libglvnd is "the GL Vendor-Neutral Dispatch library". It supports
arbitrating calls to multiple vendor libraries at runtime, and replaces
the old ways of selecting vendor libraries in modern distros.

This PR consists of a few general/improvement commits that are not directly GLVND-related, followed by the GLVND support itself in a single commit. It does not make sense to split the GLVND support further.

This will be tested in the Focal port of Ubuntu Touch. The PR is opened early to receive feedbacks from maintainers.

### CODE REVIEW

**Comment by krnlyng** [comment]
There seem to be some quite nice fixes in this, what kind of feed back do you need?

**Comment by peat-psuwit** [comment]
On the general approach and such (and if upstream wants this at all). I've successfully run a simple Qt application on it, but haven't run more complex applications. My intent is to un-draft it when I can run Lomiri with this, as the complexity should allows it to hit edge cases.

**Comment by mlehtima** [comment]
I saw a new build warning when doing test build of this:
```
egl.c: In function '_my_eglCreatePlatformWindowSurfaceEXT':
egl.c:330:45: warning: passing argument 3 of 'eglCreateWindowSurface' makes integer from pointer without a cast [-Wint-conversion]
   return eglCreateWindowSurface(dpy, config, native_window, attrib_list);
                                                                          ^~~~~~~~~~~~~
egl.c:295:23: note: expected 'EGLNativeWindowType' {aka 'long unsigned int'} but argument is of type 'void *'
    EGLNativeWindowType win,
```

File: hybris/egl/ws.c [hunk 1]
```diff
@@ -27,117 +27,135 @@
...
+ */
+EGLBoolean ws_init(const char * egl_platform)
+{
+	if (ws != NULL)
+		return ensureCorrectWs(egl_platform);
 
+	pthread_mutex_lock(&mutex);
+	if (ws != NULL) {
 		pthread_mutex_unlock(&mutex);
+		return ensureCorrectWs(egl_platform);;
```
Comment (mlehtima) [inline hybris/egl/ws.c:55]: Style nitpick, two ; at the end.

File: hybris/egl/ws.c [hunk 2]
```diff
@@ -27,117 +27,135 @@
...
 	}
+
+	char ws_lib_path[2048];
+
+	const char *eglplatform_dir = PKGLIBDIR;
+	const char *user_eglplatform_dir = getauxval(AT_SECURE)
+										? NULL
+										: getenv("HYBRIS_EGLPLATFORM_DIR");
```
Comment (mlehtima) [inline hybris/egl/ws.c:63]: Style nitpick, the indentation of these lines doesn't look good now, use the original formatting or less indentation.

**Review by mlehtima — COMMENTED** [review]
Is there some reason not to squash the last two commits which fix issues introduced in earlier commits in to the commits those are fixing?

**Comment by peat-psuwit** [comment]
> Is there some reason not to squash the last two commits which fix issues introduced in earlier commits in to the commits those are fixing?

Indeed there isn't (can't remember why I didn't). So, they're squashed now.

Changes since previous version:
- Some comment grammar fixes. A comment correction.
- Add explicit cast in `_my_eglCreatePlatformWindowSurfaceEXT()` to silence the warning.
- Remove an extra whitespace in `__eglGLVNDGetVendorString()`.
- Styling fixes.

File: hybris/egl/ws.c (hunk 1 above)
Comment (peat-psuwit) [reply hybris/egl/ws.c]: Fixed.

File: hybris/egl/ws.c (hunk 2 above)
Comment (peat-psuwit) [reply hybris/egl/ws.c]: Opps. It happens because I set my editor to render a tab character as 4 spaces. That's fixed now.

File: hybris/egl/glvnd/generate/eglFunctionList.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
+"""
```
Comment (saukko) [inline hybris/egl/glvnd/generate/eglFunctionList.py:1]: Missing license header is this also under MIT like other added .py files?

**Comment by saukko** [comment]
Just a generic note that we should add some description to the git root that states what parts are under what license, at least currently Apache 2.0 and MIT license in the git tree, didn't check if something else as well. Not mantory to fix in PR but just a note :)

File: hybris/egl/glvnd/generate/eglFunctionList.py (hunk 3 above)
Comment (peat-psuwit) [reply hybris/egl/glvnd/generate/eglFunctionList.py:1]: Probably yes; the file originally comes from Mesa. The original file doesn't have the notice either, but I think we can assume the same license as the rest of Mesa.
https://github.com/mesa3d/mesa/blob/main/src/egl/generate/eglFunctionList.py

Although, I would say that this file contains mostly data and not much code, and so might be not copyrightable. In any case, I'll update another version with the header some time soon.

File: hybris/egl/egl.c [hunk 4]
```diff
@@ -139,9 +171,69 @@ struct _EGLDisplay *hybris_egl_display_get_mapping(EGLDisplay display)
...
+
+	const char * hybris_ws;
+	switch (platform) {
+		case EGL_NONE:
+			hybris_ws = _defaultEglPlatform();
+			break;
+
+		case EGL_PLATFORM_ANDROID_KHR:
+			// "null" ws passthrough everything, which essentially means
+			// the Android platform. Beware confustion with NULL value.
```
Comment (mlehtima) [inline hybris/egl/egl.c:217]: Typo: confustion should be confusion.

File: hybris/egl/glvnd/eglglvnd.cpp [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    // Use libhybris's normal eglQueryString() otherwise.
+    return eglQueryString(dpy, name);
+}
+
+static const char *
+__eglGLVNDGetVendorString(int name)
+{
+    if (name == __EGL_VENDOR_STRING_PLATFORM_EXTENSIONS) {
+        return
+             "EGL_KHR_platform_android"
```
Comment (mlehtima) [inline hybris/egl/glvnd/eglglvnd.cpp:100]: Formatting: indentation doesn't match the lines below.

File: hybris/egl/glvnd/eglglvnd.cpp [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+static EGLBoolean __eglGLVNDgetSupportsAPI (EGLenum api)
+{
+    return api == EGL_OPENGL_ES_API;
+}
+
+static void *
+__eglGLVNDGetProcAddress(const char *procName)
+{
+    if (strcmp(procName ,"eglQueryString") == 0)
```
Comment (mlehtima) [inline hybris/egl/glvnd/eglglvnd.cpp:118]: Formatting: space should be after "," not before.

File: hybris/egl/glvnd/eglglvnd.cpp [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+ */
+
+__attribute__ ((visibility ("default")))
+EGLBoolean
+__egl_Main(uint32_t version, const __EGLapiExports *exports,
+     __EGLvendorInfo *vendor, __EGLapiImports *imports)
+{
+    if (EGL_VENDOR_ABI_GET_MAJOR_VERSION(version) !=
+        EGL_VENDOR_ABI_MAJOR_VERSION)
+       return EGL_FALSE;
```
Comment (mlehtima) [inline hybris/egl/glvnd/eglglvnd.cpp:136]: Indentation is multiples of 4 in other parts of the file. For readability maybe indent the second line in if condition with another 4.

File: hybris/egl/ws.c [hunk 8]
```diff
@@ -27,117 +27,135 @@
...
+	if (wsmod==NULL)
+	{
+		fprintf(stderr, "ERROR: %s\n\t%s\n", ws_lib_path, dlerror());
+		assert(0);
+	}
+	ws = dlsym(wsmod, "ws_module_info");
+	assert(ws != NULL);
+	ws->init_module(&hybris_egl_interface);
+
+	// Keep the name of choosen ws so that future calls can check.
```
Comment (mlehtima) [inline hybris/egl/ws.c:79]: Typo: choosen should be chosen.

**Review by mlehtima — COMMENTED** [review]
Some small typos and style issues found in the code. Otherwise looks good. Tested the non-glvnd parts and seemed to work correctly on a device.

**Comment by bigOconstant** [comment]
Any progress on getting this merged? I'm looking forward to having some hardware acceleration on my oneplus 6t

**Comment by krnlyng** [comment]
> Any progress on getting this merged? I'm looking forward to having some hardware acceleration on my oneplus 6t

A few things still need to be resolved (see above), but getting close.

File: hybris/egl/egl.c (hunk 4 above)
Comment (peat-psuwit) [reply hybris/egl/egl.c]: I've reworded it completely. There shouldn't be a typo this time, because the phrase is copied from Wikipedia's template. :D

File: hybris/egl/glvnd/eglglvnd.cpp (hunk 5 above)
Comment (peat-psuwit) [reply hybris/egl/glvnd/eglglvnd.cpp]: I've aligned the letter `E`, not the quotation mark, with the second next line. The idea was that we need a space for each token, but I don't want an additional space at the start of the string.

Anyway, have a look at Mesa's code, and they align the quotation marks. So I followed you and Mesa.

File: hybris/egl/glvnd/eglglvnd.cpp (hunk 6 above)
Comment (peat-psuwit) [reply hybris/egl/glvnd/eglglvnd.cpp]: Fixed.

File: hybris/egl/glvnd/eglglvnd.cpp (hunk 7 above)
Comment (peat-psuwit) [reply hybris/egl/glvnd/eglglvnd.cpp]: Fixed.

File: hybris/egl/ws.c (hunk 8 above)
Comment (peat-psuwit) [reply hybris/egl/ws.c]: Fixed.

**Comment by peat-psuwit** [comment]
Sorry for the delay. I forgot about this change for a while, until @krnlyng's comment pinged me.

Other than responding to the reviews:
- Update the EGL header update commit. By the time this is merged, it's probably not-so-latest anymore.
- Update the override of `eglDestroyImageKHR()`, as per the recent commit.

So, please take another look.

**Comment by MrCyjaneK** [comment]
I'm getting `undefined symbol: eglCreatePlatformWundowSurface` when launching some apps on droidian, will this pull request fix it?

**Review by mlehtima — APPROVED** [review]

**Comment by krnlyng** [comment]
@peat-psuwit i think the (c) headers are still missing for these two

https://github.com/libhybris/libhybris/blob/362efd97cd5b18f580c402d3fc14555f1ef6e31e/hybris/egl/glvnd/egldispatchstubs.c
https://github.com/libhybris/libhybris/blob/362efd97cd5b18f580c402d3fc14555f1ef6e31e/hybris/egl/glvnd/egldispatchstubs.h

otherwise i think i am ready to merge this.

**Comment by mlehtima** [comment]
Also these files are missing copyright
https://github.com/libhybris/libhybris/blob/362efd97cd5b18f580c402d3fc14555f1ef6e31e/hybris/egl/glvnd/generate/egl_hybris.xml
https://github.com/libhybris/libhybris/blob/362efd97cd5b18f580c402d3fc14555f1ef6e31e/hybris/egl/glvnd/generate/egl_other.xml

**Comment by mlehtima** [comment]
@peat-psuwit Do you have plans to add the missing copyright headers mentioned in above comments so we could get this merged?

**Comment by peat-psuwit** [comment]
Sorry, forgot about this again. Copyright headers are added in the latest commit. Please take another look.

**Review by mlehtima — APPROVED** [review]

**Comment by krnlyng** [comment]
thanks!

**Comment by peat-psuwit** [comment]
Wait a minute, why did I introduce an "advertising clause" in egldispatchstub.*?

**Comment by peat-psuwit** [comment]
Turns out, the advertising clause might cause some problem...

https://gitlab.freedesktop.org/glvnd/libglvnd/-/issues/221

**Comment by krnlyng** [comment]
@peat-psuwit seems glvnd did fix the license terms. do we need to fix something here as well?
